src/Synth/OscilGen.cpp

```diff
@@ -522,6 +513,171 @@ void OscilGen::convert2sine()
     prepare();
 }
 
+class timer_t
+{
+    using time_point = std::chrono::steady_clock::time_point;
+    std::size_t num;
+    time_point begin;
+public:
+    timer_t(std::size_t num) :
+        num(num),
+        begin(std::chrono::steady_clock::now()) {}
+    ~timer_t() {
+         time_point end = std::chrono::steady_clock::now();
+         auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(end - begin).count();
+         if(ms == 0)
+            std::cout << "Calculated " << num << " wavetables in <1 ms" << std::endl;
+         else
+            std::cout << "Calculated " << num << " wavetables in " << ms << " ms" << std::endl;
+    }
+};
+
+wavetable_types::WtMode OscilGen::calculateWaveTableMode(bool forceWtMode) const
+{
+    using WtMode = wavetable_types::WtMode;
+    if(forceWtMode)
+    {
+        return WtMode::freqwave_smps;
+    }
+    else
+    {
+        return mayUseRandom() ? WtMode::freqseed_smps : WtMode::freq_smps;
+    }
+}
+
+std::size_t OscilGen::calculateNumFreqs(bool voice_uses_reso) const
+{
+    // Calculate required size for freq scale tensor
+    // For Resonance, we need this for every key (because the Resonance can
+    // differ a lot between only two keys),
+    // otherwise use some sane default
+    return (voice_uses_reso && res && res->Penabled) ? 128 : WaveTable::num_freqs;
```

Comment:
Medium priority note as a memory saving incentive num_freqs can be lower (assuming resonance is disabled), but on the other hand WaveTable::num_freqs is only 10, so maybe that isn't that big of a deal.

Specifically I'm thinking of memory at defaults vs when we're actually forced to use worst case memory footprints. e.g. for the basic sine wave you'd only need one wavetable and for other waveshapes you only need additional frequencies when the aliasing threshold starts to modify things.

Perhaps a viable workaround (likely out of scope for the base PR) would be to add in a numeric similarity check with generated wavetables before passing them over to the RT side. Given a sufficiently low L2 distance between two wavetables it ought to be safe to cull waveforms and that is easier to code than understanding the expected output based on all of zyn's parameters prior to wavetable creation.

Reply:
Hmm maybe "10" is indeed not the best. It might be best to do a divide and conquer algorithm and as soon as you discover that the distance L2 distance between two waves is small enough, you stop the recursion. This would even work with resonance enabled. The result could be a wave table where the waves have different density for different frequency ranges. Does that make sense?

Reply:
Yeah, that's the sort of idea I was thinking about. Initially I was thinking

Step 1) calculate a bunch of waves
Step 2) realize a bunch of different frequencies yield effectively the same result
Step 3) discard waves

The challenge with making it a more complex process is parallelism. For moderate core counts it's faster to calculate and then throw away than it is to conditionally calculate things depending on other things, but that's just another byproduct. My main focus was avoiding excess memory thinking primarily of usecases like the zynthian setup.